### PR TITLE
Exclusivity refactor and fixes.

### PR DESCRIPTION
**Pull request by atrick**
Refactor exclusivity logic in preparation for exhaustive round-trip verification.

This also makes the code more robust and enables a fix for <rdar://36989789> [SR-6862]: "Unexpected partial_apply use" in SILOptimizer/Mandatory/DiagnoseStaticExclusivity.cpp.

### CODE REVIEW

**Comment by atrick** [comment]
@swift-ci test.

**Comment by atrick** [comment]
@swift-ci Test Source Compatibility

**Comment by atrick** [comment]
@swift-ci Test Source Compatibility

File: lib/SILOptimizer/Mandatory/DiagnoseStaticExclusivity.cpp [hunk 1]
```diff
@@ -1038,55 +936,37 @@ static void checkForViolationsInNoEscapeClosureArguments(
...
-    if (Callee->isThunk() == IsReabstractionThunk) {
-      // For source compatibility reasons, treat conflicts found by
-      // looking through reabstraction thunks as warnings. A future compiler
-      // will upgrade these to errors;
-      bool WarnOnThunkConflict = true;
-      // Recursively check any arguments to the partial apply that are
-      // themselves noescape closures. This detects violations when a noescape
-      // closure is captured by a reabstraction thunk which is itself then passed
-      // to an apply.
-      checkForViolationsInNoEscapeClosureArguments(Accesses, PAI, ASA,
```
Comment (devincoughlin) [inline lib/SILOptimizer/Mandatory/DiagnoseStaticExclusivity.cpp]: I particularly like that you were able to remove the recursion here.

**Review by devincoughlin — APPROVED** [review]
Looks good to me. Much cleaner!

**Pull request merged**